File: apps/agent/src/agents/adaptation.ts
import { ChatPromptTemplate, MessagesPlaceholder } from '@langchain/core/prompts';
import { RunnableSequence } from '@langchain/core/runnables';
import { AgentType, ActionItem } from '../types';
import { createLLM, getSystemPrompt } from '../utils/llm';
import { StructuredOutputParser } from 'langchain/output_parsers';
import { z } from 'zod';
import { AIMessage } from '@langchain/core/messages';
import { StateAnnotation } from '../types';

// Process the user input with Adaptation agent
export const processAdaptation = async (state: typeof StateAnnotation.State): Promise<typeof StateAnnotation.State> => {
  // Create LLM
  const llm = createLLM('gpt-4o', 0.4); // Higher temperature for creative adaptations

  // Create parser for structured output
  const outputParser = StructuredOutputParser.fromZodSchema(
    z.object({
      actions: z.array(
        z.object({
          type: z.enum([
            'updateTask',
            'updateManyTasks',
            'logActivity',
            'none'
          ]),
          payload: z.any()
        })
      ),
      adaptationStrategy: z.string().describe('The strategy you recommend for adapting the task or plan'),
      reasoning: z.string().describe('Your explanation of why adaptation is needed and how it will help'),
      response: z.string().describe('A concise, helpful response to the user explaining the adaptation strategy and changes')
    })
  );

  // Prepare the conversation history
  const conversationHistory = state.messages.filter(msg => 
    msg.getType() === 'human' ||
    (msg.getType() === 'ai' && msg.additional_kwargs.agentType === AgentType.Adaptation)
  );

  // Task and user context
  const taskContext = state.task ? 
    `Task: ${state.task.title}\nDescription: ${state.task.description || 'None'}\nPriority: ${state.task.priority}\nStage: ${state.task.stage}\nStatus: ${state.task.stageStatus}\nDeadline: ${state.task.deadline ? new Date(state.task.deadline).toISOString() : 'None'}` : 
    'No task provided';

  // Create a prompt template
  const prompt = ChatPromptTemplate.fromMessages([
    ['system', getSystemPrompt('adaptation') + `\n\nRespond with a structured output containing actions, an adaptation strategy, reasoning, and a concise user-friendly response.`],
    new MessagesPlaceholder('conversation_history'),
    ['human', `User request: {input}\n\nTask Context:\n${taskContext}\n\nAnalyze if the current task/plan needs adaptation. Consider if the approach should be modified, if timelines need adjustment, or if goals need to be recalibrated. Provide a structured response with actions to take in JSON format. Include a concise, helpful response to the user explaining the adaptation strategy and changes. {format_instructions}`],
  ]);

  // Create the chain
  const chain = RunnableSequence.from([
    {
      input: (state: typeof StateAnnotation.State) => state.input,
      conversation_history: (state: typeof StateAnnotation.State) => conversationHistory,
      format_instructions: async () => outputParser.getFormatInstructions()
    },
    prompt,
    llm,
    outputParser
  ]);

  // Execute the chain
  const result = await chain.invoke(state);

  // Record the agent's adaptation strategy as a message
  if (result.adaptationStrategy) {
    state.messages.push(new AIMessage({
      content: result.adaptationStrategy,
      additional_kwargs: {
        agentType: AgentType.Adaptation,
        name: 'adaptation'
      }
    }));
  }

  // Record the agent's thought process as a message
  if (result.reasoning) {
    state.messages.push(new AIMessage({
      content: result.reasoning,
      additional_kwargs: {
        agentType: AgentType.Adaptation,
        name: 'reasoning'
      }
    }));
  }

  // Store the actions for execution
  state.actionItems = result.actions.filter((action) => action.type !== 'none') as ActionItem[];
  
  // Create response for the user
  if (result.response) {
    state.agentResponse = result.response;
    
    // Push the agent's response to the message history
    state.messages.push(
      new AIMessage({
        content: state.agentResponse,
        additional_kwargs: {
          agentType: AgentType.Adaptation,
        },
      })
    );
  }
  
  return state;
};

File: apps/agent-python/src/agents/adaptation.py
from typing import Dict, Any
from langchain_core.prompts import ChatPromptTemplate, MessagesPlaceholder
from langchain_core.messages import AIMessage
from langchain.output_parsers import StructuredOutputParser, ResponseSchema
from ..types import State, AgentType, ActionItem
from ..utils.llm import create_llm, get_system_prompt


async def process_adaptation(state: State) -> Dict[str, Any]:
    """
    Process the user input with Adaptation agent.

    Args:
        state: Current state

    Returns:
        Dictionary of state updates
    """
    # Create LLM with higher temperature for creative adaptations
    llm = create_llm('gpt-4o', 0.4)

    # Create parser for structured output
    response_schemas = [
        ResponseSchema(
            name="actions",
            description="Array of actions: updateTask, updateManyTasks, logActivity, none"
        ),
        ResponseSchema(
            name="adaptationStrategy",
            description="The strategy you recommend for adapting the task or plan"
        ),
        ResponseSchema(
            name="reasoning",
            description="Your explanation of why adaptation is needed and how it will help"
        ),
        ResponseSchema(
            name="response",
            description="A concise, helpful response to the user explaining the adaptation strategy and changes"
        )
    ]
    output_parser = StructuredOutputParser.from_response_schemas(response_schemas)

    # Prepare conversation history
    conversation_history = [
        msg for msg in state.messages
        if msg.type == 'human' or (
            msg.type == 'ai' and
            hasattr(msg, 'additional_kwargs') and
            msg.additional_kwargs.get('agentType') == AgentType.ADAPTATION.value
        )
    ]

    # Task context
    task_context = "No task provided"
    if state.task:
        task_context = f"""Task: {state.task.title}
Description: {state.task.description or 'None'}
Priority: {state.task.priority}
Stage: {state.task.stage or 'None'}
Status: {state.task.stageStatus or 'None'}
Deadline: {state.task.deadline or 'None'}"""

    # Create prompt template
    prompt = ChatPromptTemplate.from_messages([
        ('system', get_system_prompt('adaptation') + """

Respond with a structured output containing actions, an adaptation strategy, reasoning, and a concise user-friendly response."""),
        MessagesPlaceholder(variable_name='conversation_history'),
        ('human', f"""User request: {{input}}

Task Context:
{task_context}

Analyze if the current task/plan needs adaptation. Consider if the approach should be modified, if timelines need adjustment, or if goals need to be recalibrated. Provide a structured response with actions to take in JSON format. Include a concise, helpful response to the user explaining the adaptation strategy and changes.

{{format_instructions}}""")
    ])

    # Create the chain
    chain = prompt | llm | output_parser

    # Execute the chain
    result = await chain.ainvoke({
        'input': state.input,
        'conversation_history': conversation_history,
        'format_instructions': output_parser.get_format_instructions()
    })

    updates = {'messages': []}

    # Record adaptation strategy
    if result.get('adaptationStrategy'):
        updates['messages'].append(AIMessage(
            content=result['adaptationStrategy'],
            additional_kwargs={
                'agentType': AgentType.ADAPTATION.value,
                'name': 'adaptation'
            }
        ))

    # Record reasoning
    if result.get('reasoning'):
        updates['messages'].append(AIMessage(
            content=result['reasoning'],
            additional_kwargs={
                'agentType': AgentType.ADAPTATION.value,
                'name': 'reasoning'
            }
        ))

    # Store the actions
    actions = result.get('actions', [])
    if isinstance(actions, list):
        action_items = []
        for action in actions:
            # Skip if action is a string or not a dict
            if not isinstance(action, dict):
                continue

            action_type = action.get('type')
            if action_type and action_type != 'none':
                payload = action.get('payload')
                # Ensure payload is a dict or None
                if payload is not None and not isinstance(payload, dict):
                    payload = None
                action_items.append(ActionItem(type=action_type, payload=payload))

        updates['actionItems'] = action_items

    # Create response
    if result.get('response'):
        updates['agentResponse'] = result['response']
        updates['messages'].append(AIMessage(
            content=result['response'],
            additional_kwargs={'agentType': AgentType.ADAPTATION.value}
        ))

    return updates
